Set-up initial documentation tree and first "Integration" docs

gdemonet opened the pull request:
We define initial documentation for "integrating with MetalK8s", a.k.a. Solutions.

To build the docs, you can use:

```
# Will build HTML files under docs/_build
tox -e docs

# Will watch docs/ sources and serve built HTML files locally on port 8000
tox -e docs -- livehtml
```

Please make sure to read through and discuss anything feeling unclear. Details regarding the integration with Operators and their deployment will be added in the future, during the investigation work on #1060 and #1061.

**Summary**:

- [x] Define a base "integration" document for external solutions
- [x] Define Solutions archiving guidelines
- [x] Define Solutions Operator guidelines

_Note: UI guidelines are not done in this PR_

---

Fixes: #951
Fixes: #952
Fixes: #953

bert-e commented:
# Hello gdemonet,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/metalk8s/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `feature/951-solutions-base-docs` into
`development/2.0`
* [w/2.1/feature/951-solutions-base-docs](https://github.com/scality/metalk8s/compare/development/2.1...w/2.1/feature/951-solutions-base-docs) will be merged into `development/2.1` 
* [w/2.2/feature/951-solutions-base-docs](https://github.com/scality/metalk8s/compare/development/2.2...w/2.2/feature/951-solutions-base-docs) will be merged into `development/2.2` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @cmonfort

* @alexandre-allard-scality

* @gdemonet

* @slaperche-scality

* @ramanduh

* @ballot-scality

* @teddyandrieux

* @tcarmet

* @lucieleonard

* @zempashi

* @sayf-eddine-scality

* @carlito-scality

* @jbertran

* @thomasdanan

* @monpote

* @nootal

* @nicolast

* @alxf

bert-e commented:
# History mismatch



Merge commit #d3b281900c8927ea5caf1981e3c2751e555daf92 on the integration branch
`w/2.1/feature/951-solutions-base-docs` is merging a branch which is neither the current
branch `feature/951-solutions-base-docs` nor the development branch
`development/2.1`.

It is likely due to a rebase of the branch `feature/951-solutions-base-docs` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

gdemonet committed:
Introduce Solution "integration" documentation

Issue: GH-951

gdemonet committed:
Define Solution archive guidelines

Issue: GH-952

gdemonet committed:
Define Solution Operator guidelines

Issue: GH-953

gdemonet committed:
Add plantuml to docs/requirements

rachedbenmustapha commented on docs/developer/solutions/archive.rst:
```diff
@@ -0,0 +1,152 @@
+
+.. note::
+
+   Each Solution ISO will be made available to Pods as a single repository.
```
I don't really understand this paragraph. What I came to understand is that all images listed in all instances of `ImagesList` CR in a Solution will be sort of namespaced in a common "virtual" image registry repository (either by having a cluster-local registry, or by using unique prefixes when injecting the images in node's `containerd`s).

If that is the case then it answers a question that I had: what is the expected behavior if several solutions embed conflicting tags for the same image, for example `latest` or `8-latest`. Even if it is a practice that should be discouraged, it could still happen. Documenting the expected behavior (whether it's a problem or a non-problem) in case of conflict would be useful.

gdemonet replied:
> either by having a cluster-local registry, or by using unique prefixes when injecting the images in node's `containerd`s

There will be a cluster-local _registry_, which role will be to expose all images defined in `ImagesList` under a single _repository_. Let's clarify:

1. Assume I have a "Example v1.0" solution packaged, in which I have a `example:1-latest` image
2. After importing this Solution, MetalK8s will expose a new repository at a URL like `<registry_url>/example-1.0/`
3. The "Example v1.0 Operator" will then rely on this repository prefix to reference the image, so when creating an "Example" instance, use `image: <registry_url>/example-1.0/example:1-latest`

Now, let's assume I want to upgrade to "Example v1.1", which still include the `example:1-latest` image.

1. After importing this new version, I have a new repository at `<registry_url>/example-1.1/`
2. The existing Operator can thus reference either of those images, based on his own logic, still using the `example:1-latest` tag

So there would be no conflict w.r.t. tags. However, how does the Operator know which repository to use? See comment below for more details.

rachedbenmustapha replied:
It's a lot clearer when explained this way :) I figured when looking at referenced issues/PRs but what got me was "what do ISOs have to do with Pods?"

gdemonet replied:
I explained slightly more in the docs, though this will be detailed in the future with examples and more diagrams.

rachedbenmustapha commented on docs/developer/solutions/archive.rst:
```diff
@@ -0,0 +1,152 @@
+   kind: ImagesList
+   solution: MySolution
+   images:
+     - archive: my_image_1_0.tar.gz
```
This format seems to actively discourage using `docker save` of multiple images at once. While this is perfectly fine, are we 100% set on disallowing it? Maybe it has changed since, but from my experience saving multiple images at once does efficient layer deduplication and gives better size after compression than compressing each image layer graph individually. The size and time benefits may not be worth the schema simplicity compromise, just wondering.

NicolasT replied:
This part is outdated, we're currently integrating another approach to serve container images into MetalK8s (statically, using nothing but `nginx`), and 'solutions' shall use the same approach. This is *not* based on `docker save` but different tooling, and does allow for efficient deduplication (which, indeed, the approach currently described does not) by using `hardlink`(s).

gdemonet replied:
Thanks for the clear answer @NicolasT . Indeed, this is mentioned in the [`todo` above](https://github.com/scality/metalk8s/pull/1020/files#diff-0c567a23053ae27cf85c625171c5d19aR97). Let's track this topic in #1049

gdemonet replied:
I reworked these guidelines to match with what was introduced in #1049

rachedbenmustapha commented on docs/developer/solutions/archive.rst:
```diff
@@ -0,0 +1,152 @@
+
+   Each Solution ISO will be made available to Pods as a single repository.
+   Operator Deployments will then be required to accept the repository prefix,
+   stored in a ``ConfigMap``, in order to properly configure the resources they
```
Does it mean that only select tailor-made operators with this `ConfigMap` support will work, and that one can't (at least yet) package an existing operator as a solution? May be worth a note if that's the case.

gdemonet replied:
Operators need to know _where_ the images exist, so they can reference them in the various objects they create. What we suggest here is that we will define a `ConfigMap` to mount as an environment variable, for Operators to use as their `REPOSITORY_PREFIX`.

Let's take our "Example" solution described in the comment above:

- I have imported "Example v1.0", images are available in some repository
- I retrieve the full URL of the Operator image, including the repository (_Note: this is cumbersome, we may want to find a better way_)
- I deploy the Operator for this version, passing into the container spec:

```
image: {{ full_operator_image_url }}
[...]
env:
- name: REPOSITORY_PREFIX
    valueFrom:
    configMapKeyRef:
        name: metalk8s-solutions-repositories
        key: example-1.0
```

- The Operator container can now use this to generate its own resources, using `image: {{ REPOSITORY_PREFIX }}/example:1-latest`

Now, suppose (as in the comment above) that I have deployed version 1.1 of my Example solution. If I want to upgrade the 1.0 Operator, I can edit its Deployment, to use the new Operator image, and pass change the ConfigMap key to use, which will effectively upgrade all instances of `example:1-latest` deployed.

I know this all feels complicated, and thanks a lot for taking the time to review that part. We'll clarify these processes as soon as we know more (working on a _Hello-world Solution_ in the coming weeks should help us on that).

rachedbenmustapha replied:
Once each operator knows where to get images it's pretty straightforward and it's good that it's "managed" by the deployer that way.

> I retrieve the full URL of the Operator image, including the repository (Note: this is cumbersome, we may want to find a better way)

For bootstrapping each Operator and making the initial manifests configurable, any opinions on `kustomize` / `kubectl apply -k`? It can be used [as an API](https://github.com/kubernetes-sigs/kustomize/blob/master/pkg/types/kustomization.go#L42) and allows replacing images (in addition to general patching) pretty easily as input data, leaving operator deployment manifests "clean". I can see how it only shifts one part of the problem elsewhere and does not entirely solve it though :)

gdemonet replied:
I'll look into `kustomize`, thanks for the suggestion :)

gdemonet replied:
So, I looked into `kustomize`, it's really interesting. It doesn't seem to be available as part of `apiserver`, only inside `kubectl`, but it could be a good addition to the manifests provided in the ISO. Let's wait until we know more with #1060, I'll try it out.

I added a note to mention this in the docs.

rachedbenmustapha commented on docs/developer/solutions/introduction.rst:
```diff
@@ -0,0 +1,115 @@
+
+**MAY**:
+
+- Use MetalK8s monitoring services (Prometheus and Grafana)
```
It may be worth taking a stronger stance on this. As a potential Solution developer I understand that I can use MetalK8s monitoring services for convenience, but I am not sure if it's a good (ie sustainable) idea, or if the more sustainable path is to embed my own prometheus stack.

gdemonet replied:
I thought there was no strong requirement, so moved it to _MAY_, but indeed it makes sense to force the usage of the cluster-provided Prometheus stack.

rachedbenmustapha replied:
I don't have a strong opinion either way, it's mostly about nudging developers in a "preferred" direction so that resources/storage capacity planning for the platform's Prometheus instance are adequate. As a Solution developer I would enjoy not having to manage my own Prom stack if I can trust that I won't have any surprises with retention duration etc :)

gdemonet replied:
Changed `MAY` into `SHOULD`, although I don't want to say `MUST`, since deploying your own monitoring stack should not prevent the Solution from working (though I'm not sure if having multiple Prometheus Operators watching for `ServiceMonitor`s could pose issues...). I'll open a tracking ticket.

gdemonet replied:
Ok, latest conversations led to a simpler stance: Solutions that want to be deployed on MetalK8s `MUST` use the provided Prometheus Operator to set-up their monitoring. However, it is reasonable to assume Solution developers may want their Operator to work in any K8s cluster, hence could manage their own monitoring stack there. I'm still unsure about how they would do so, let's focus on _using_ MetalK8s stack for now (will consider the alternative use-case in #1158).

rachedbenmustapha commented on docs/developer/solutions/archive.rst:
```diff
@@ -0,0 +1,152 @@
+
+   NAME=${YOUR_SOLUTION_NAME}
+   VERSION=${YOUR_PROJECT_FULL_VERSION}
+
```
This could be a subject that we explicitly don't want to touch, but what about dependencies? While the average Solution could ideally be self-contained, some Solutions could have a need to depend on an already-deployed Solution and be non-functional without it. Examples of dependencies could be:
- A low-level storage backend that would be complex-enough to manage to warrant having its own Solution
- Some high-overhead services like a Kafka+ZooKeeper stack, or ELK stack, that would be best deployed once instead of once per solution that needs them

gdemonet replied:
I agree that requesting external services / Solutions from a Solution has valid use-cases, and we may want to invest time looking at possible ways to manage this. I believe it could become a part of some "Operator of Operators" that we'd like to add in the future, for instance something like OLM from CoreOS.

I'll add a note in these docs, but I don't have a clear answer right now.

However, regarding common services, such as Kafka+Zookeeper or ELK, I believe those should be part of the original offering from MetalK8s. Thoughts @NicolasT?

rachedbenmustapha replied:
Dependency handling is a big subject so it's ok as just a note for now.

If there are services to be provided by the platform, that the Solution expects (which ELK and Prometheus already are), maybe a Solution could mention it in its own manifest, as a capability name (for optional services, should that happen) or a minimum MetalK8s version (for services that will always be provided once introduced to the platform).

gdemonet replied:
I added a `REQUIRE_METALK8S` variable in the expected `product.txt`, which will allow for a minimal dependency management. Also added a note for handling more specific dependencies, such as monitoring components, TBD later.

rachedbenmustapha commented on docs/developer/solutions/introduction.rst:
```diff
@@ -0,0 +1,115 @@
+--------------------
+
+We include a detailed interaction sequence diagram for describing how MetalK8s
+will handle user input when deploying / upgrading Solutions.
```
Would be worth mentioning rollback (even if it's exactly the same as upgrading) for operations-minded readers.

gdemonet replied:
We will add more in-depth guides for operational matters w.r.t Solutions, this one is for developers to understand how MetalK8s manages Solutions (which needs improvements as well, granted).

Most importantly, this diagram shows how deployment/upgrade is actually left to the user (manually or through the Solution UI). Rollback would fall into this category as well, IMHO.

Agreed though, we will need to show how a Solution gets **removed** from MetalK8s (which we won't do in the early stages, would be implemented in the generic `import_all_solutions` runner).

gdemonet replied:
I added a warning for removal, and mentioned more details about how an Operator is deployed manually by the user.

rachedbenmustapha commented on docs/developer/solutions/introduction.rst:
```diff
@@ -0,0 +1,115 @@
+    solutions:
+      - /solutions/storage_1.0.0.iso
+      - /solutions/storage_latest.iso
+      - /other_solutions/computing.iso
```
This example has the advantage of showing 2 versions of the same Solution, I was going to be ask about this case. It leaves me wondering how this case would work: is it designed for parallel run (either because 2 different versions will be needed indefinitely because of independent release tracks, or because we want to support gradual migration from version `v` to version `v+1`) or only one version active at a time that need to be selected/switched by the user through API/UI?

gdemonet replied:
It's designed to run multiple versions of a Solution at the same time, using Namespaces for isolation. This allows users to try a new version in some `staging` namespace first, before upgrading their `prod` namespace.

gdemonet replied:
I improved a few sections to be more explicit about this intent, later improvements will include UML diagrams for these use-cases.

rachedbenmustapha reviewed (commented):
First thoughts

charlesprost commented on docs/developer/solutions/introduction.rst:
```diff
@@ -0,0 +1,115 @@
+archived as an ISO disk image, containing:
+
+- A set of OCI images to inject in MetalK8s image registry
+- An `Operator`_, to deploy on the cluster
```
```suggestion
- An `Operator`_ to deploy on the cluster
```

charlesprost commented on docs/developer/solutions/introduction.rst:
```diff
@@ -0,0 +1,115 @@
+      - /solutions/storage_latest.iso
+      - /other_solutions/computing.iso
+
+There would be no explicit information about what the archives contain.
```
```suggestion
There would be no explicit information about what an archive contains.
```

Since we use one archive in the line just after.

charlesprost commented on docs/developer/solutions/introduction.rst:
```diff
@@ -0,0 +1,115 @@
+**MUST**:
+
+- Comply with the standard archive structure defined by MetalK8s
+- Expose management of its Operator instances in its UI
```
A few sections above we state that the UI is optional for a solution, and in here we say that it MUST expose stuff in its UI.

gdemonet replied:
Would that be better?

```suggestion
- If providing a UI, expose management of its Operator instances
```

charlesprost commented on docs/developer/solutions/operator.rst:
```diff
@@ -0,0 +1,84 @@
+.. todo::
+
+   Specify each of them, include example (after `#1060`_ is done).
+   Remember to note specifities about ``OCI_REPOSITORY_PREFIX`` / namespaces.
```
```suggestion
   Remember to note specificities about ``OCI_REPOSITORY_PREFIX`` / namespaces.
```

bert-e commented:
# Branches have diverged



This pull request's source branch `feature/951-solutions-base-docs` has diverged from
`development/2.0` by more than 50 commits.

To avoid any integration risks, please re-synchronize them using one of the
following solutions:

* Merge `origin/development/2.0` into `feature/951-solutions-base-docs`
* Rebase `feature/951-solutions-base-docs` onto `origin/development/2.0`

Note: If you choose to rebase, you may have to ask me to rebuild
integration branches using the `reset` command.

bert-e commented:
# History mismatch



Merge commit #d3b281900c8927ea5caf1981e3c2751e555daf92 on the integration branch
`w/2.1/feature/951-solutions-base-docs` is merging a branch which is neither the current
branch `feature/951-solutions-base-docs` nor the development branch
`development/2.1`.

It is likely due to a rebase of the branch `feature/951-solutions-base-docs` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

gdemonet committed:
Rework archive guidelines

Since #1049 was introduced, we have a new approach for serving OCI
images in the cluster.
We glaze over the details in this doc, and link to the
`static-container-registry` project.

Issue: #953

gdemonet committed:
Mention dependency management

We want to allow Solutions to require a specific/minimal version of
MetalK8s on which they are being deployed.

Handling of other dependencies, such as monitoring components, is not
handled yet.

Issue: #952

gdemonet committed:
Improve description of monitoring

Usage of the "kube-prometheus" deployed by MetalK8s was not documented
enough.

Issue: #951

gdemonet committed:
Improve Operator deployment explanation

Better say too much than be unclear.

Issue: #953

gdemonet commented:
@bert-e reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `feature/951-solutions-base-docs` into
`development/2.0`
* [w/2.1/feature/951-solutions-base-docs](https://github.com/scality/metalk8s/compare/development/2.1...w/2.1/feature/951-solutions-base-docs) will be merged into `development/2.1` 
* [w/2.2/feature/951-solutions-base-docs](https://github.com/scality/metalk8s/compare/development/2.2...w/2.2/feature/951-solutions-base-docs) will be merged into `development/2.2` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @cmonfort

* @jbertran

* @monpote

* @alxf

* @gdemonet

* @lucieleonard

* @zempashi

* @carlito-scality

* @alexandre-allard-scality

* @nootal

* @nicolast

* @thomasdanan

* @tcarmet

* @slaperche-scality

* @teddyandrieux

* @sayf-eddine-scality

* @ramanduh

bert-e commented:
# History mismatch



Merge commit #22c4cdfc05d4011fa494fcff5070198c01cd7973 on the integration branch
`w/2.1/feature/951-solutions-base-docs` is merging a branch which is neither the current
branch `feature/951-solutions-base-docs` nor the development branch
`development/2.1`.

It is likely due to a rebase of the branch `feature/951-solutions-base-docs` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

gdemonet commented:
@bert-e reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `feature/951-solutions-base-docs` into
`development/2.0`
* [w/2.1/feature/951-solutions-base-docs](https://github.com/scality/metalk8s/compare/development/2.1...w/2.1/feature/951-solutions-base-docs) will be merged into `development/2.1` 
* [w/2.2/feature/951-solutions-base-docs](https://github.com/scality/metalk8s/compare/development/2.2...w/2.2/feature/951-solutions-base-docs) will be merged into `development/2.2` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @cmonfort

* @jbertran

* @monpote

* @alxf

* @gdemonet

* @lucieleonard

* @zempashi

* @carlito-scality

* @alexandre-allard-scality

* @nootal

* @nicolast

* @thomasdanan

* @tcarmet

* @slaperche-scality

* @teddyandrieux

* @sayf-eddine-scality

* @ramanduh

slaperche-scality commented on docs/developer/solutions/operator.rst:
```diff
@@ -63,6 +63,12 @@ MetalK8s does not handle the monitoring of a Solution application, which means:
 
+The `Prometheus Operator`_ deployed by MetalK8s will have cluster-scoped
+permissions, and will be able to read the aforementioned ``ServiceMonitor``
+objects to set-up monitoring of your application services.
```
s/set-up/setup/

>  Setup vs. Set-Up vs. Set Up
>
> The noun setup is usually styled as a solid compound (that is, as a single word) in American English and as a hyphenated compound (set-up) in British English. The noun may be used attributively (that is, as an adjective) in such phrases as setup man (“someone who makes machine tool adjustments”) and setup tool (“software that facilitates the installation of a computer program”). The verb set up, on the other hand, is usually found as an open compound (two words, no hyphen) in both American and British English.

gdemonet replied:
Mmh from what you quoted, it would be `set up`, no? Thanks for the info btw :)

slaperche-scality replied:
Yup indeed

slaperche-scality commented on docs/developer/solutions/archive.rst:
```diff
@@ -85,8 +85,16 @@ General product information about the packaged Solution must be stored in the
 
-   NAME=${YOUR_SOLUTION_NAME}
-   VERSION=${YOUR_PROJECT_FULL_VERSION}
+   NAME=Example
```
in commit message "Solutions  to": two many spaces beteen the two words.

gdemonet replied:
Done

slaperche-scality commented on docs/developer/solutions/archive.rst:
```diff
@@ -94,51 +98,64 @@ It must respect the following format::
 
-.. todo::
+MetalK8s exposes container images in the OCI_ format through a static
+read-only registry. This registry is built with NGINX_, and relies on having
```
minor: s/NGINX/nginx/g

I have the impression that the uppercase is only used for the name of the company and the logo, when talking about the sofware it's almost always written as `nginx`.

gdemonet replied:
OK, wasn't sure which one to use

slaperche-scality commented on docs/developer/solutions/archive.rst:
```diff
@@ -94,51 +98,64 @@ It must respect the following format::
 
-.. note::
+The images names and tags will be inferred from the directory names chosen when
+using ``skopeo copy``. Using `hardlink` is highly recommended if one wants to
```
>  Using `hardlink` is highly recommended if one wants to define alias tags for a single image.

I'm not sure to understand this part. IIUC, hardlinks are used to deduplicate the layers, whereas aliases are provided by some `nginx` config.

gdemonet replied:
Will the procedure above handle multiple tags for a single image? I thought tags were just derived from directories. Hence, if you want two tags for a single image, you would `skope copy` it twice, then `hardlink` to remove all duplicate layers.

slaperche-scality replied:
Indeed, `nginx` aliases are something else and not related on what you were saying here. I got confused.

slaperche-scality commented on docs/developer/solutions/archive.rst:
```diff
@@ -0,0 +1,152 @@
+and Joliet_ directory records. The character encoding must be UTF-8_. The
+conformance level is expected to be at most 3, meaning:
+
+- Directory identifiers may not exceed 31 characters in length
```
What's a characters here: A byte (I suppose)? A Unicode codepoint? A grapheme cluster (highly unlikely)?

Since we're allowing/requiring UTF-8 here, a character is something ambiguous and may not match what people expect.

gdemonet replied:
Nothing clear from what I read so far, feel free to go through the [specs](http://www.ecma-international.org/publications/standards/Ecma-119.htm) ;). In the meantime, I'll mention `bytes`.

gdemonet replied:
TBH, I just read through the docs and tried to translate the command we use in the buildchain into some text, but I have no clue if what I wrote is correct (we probably don't care much if we can mount and read from the ISO).

slaperche-scality replied:
IIUC, ISO 9660 doesn't support Unicode characters so there is no issue (one character == one byte, hence the restrictions expressed in term of character).
Joliet use UCS-2 encoding (which doesn't support all the Unicode characters).

I guess that the UTF-8 is supported through Rock Ridge extensions.

Seems complicated, so let's use mention `bytes` as you suggest :D

slaperche-scality commented on docs/developer/solutions/archive.rst:
```diff
@@ -0,0 +1,152 @@
+.. _Joliet: https://en.wikipedia.org/wiki/Joliet_(file_system)
+.. _UTF-8: https://tools.ietf.org/html/rfc3629
+
+Here is an example invocation of the common Unix mkisofs_ tool to generate such
```
such an archive => such archive?

bert-e commented:
# History mismatch



Merge commit #1ee819b12300829d8fbbc4fb3c50cf5c0709d83f on the integration branch
`w/2.1/feature/951-solutions-base-docs` is merging a branch which is neither the current
branch `feature/951-solutions-base-docs` nor the development branch
`development/2.1`.

It is likely due to a rebase of the branch `feature/951-solutions-base-docs` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

NicolasT commented on docs/developer/solutions/archive.rst:
```diff
@@ -0,0 +1,177 @@
+
+   NAME=Example
+   VERSION=1.0.0-dev
+   REQUIRE_METALK8S=">=2.0"
```
Do we want to require the inclusion of `GIT`, `BUILD_TIMESTAMP` as done in MetalK8s proper?

gdemonet replied:
We can mention them as useful/recommended, but I don't think our tooling will use them, will it?

NicolasT replied:
Not necessarily indeed.

NicolasT commented on docs/developer/solutions/archive.rst:
```diff
@@ -0,0 +1,177 @@
+A detailed procedure for generating the expected layout is available at
+`NicolasT/static-container-registry`_. You can use the script provided there
+to generate your own NGINX configuration and test this static registry for
+yourself. MetalK8s will generate its own configuration when importing the
```
> MetalK8s will generate its own configuration when importing the Solution archive.

Is that so? In this case, we'll need to ship `static-container-registry` with MetalK8s... For MetalK8s proper, we *don't* generate the configuration at runtime, but at build-time...

gdemonet replied:
Yes, I don't see how Solutions could include their own without knowing where the ISO will get mounted in our `repositories` container. But maybe you have an idea?

NicolasT replied:
That's why the storage path can be any random string, including... a variable, as explained in the `static-container-registry` `README`, and as used by MetalK8s as well.

NicolasT commented on docs/developer/solutions/archive.rst:
```diff
@@ -0,0 +1,177 @@
+
+   Operator Deployments should not rely on this naming pattern for finding
+   the images for their resources. Instead, the full repository prefix will be
+   stored in a ``ConfigMap``, that Deployments will be able to expose as
```
Why should it be in a `ConfigMap`? That seems cumbersome?

gdemonet replied:
Because it needs to be readable somewhere that is not Salt, if users want to know where their images are stored. I thought `ConfigMap`s could help provide a consistent storage destination such that mounting its value as an environment variable in Operator Deployments is merely about knowing the `ConfigMap` name and the key for this Solution.

Or do we know the naming of repositories will be stable/predictable and only need to pass the registry address (in which case we probably just need to read the right `Service` object)?

NicolasT replied:
Fair enough, though I don't think solutions should be aware of this `ConfigMap`, or refer to it in their deployment manifest. That's cumbersome. We can provide or inject that value when rendering the manifest directly.

NicolasT commented on docs/developer/solutions/archive.rst:
```diff
@@ -0,0 +1,177 @@
+using ``skopeo copy``. Using `hardlink` is highly recommended if one wants to
+define alias tags for a single image.
+
+MetalK8s also requires that container images respect specific standards, such
```
Does MetalK8s require this? :thinking:

gdemonet replied:
Dunno, we have a section in `requirements` that mentions OCI standards, I thought Solutions would fall in the same category. I can remove this section if you want.

NicolasT replied:
Merely referring to 'using the same base image' and having specific **LABEL**s. That's not required.

NicolasT commented on docs/developer/solutions/deployment.uml:
```diff
@@ -0,0 +1,89 @@
+user -> bootstrap : Add ISO path to the "SolutionsConfiguration" file
+user -> saltmaster ++ : Request Solutions import
+
+saltmaster <-> apiserver : Retrieve "MetalK8sSolutions" ConfigMap
```
In general, object names are not CamelCased in K8s-land.

gdemonet replied:
Oops, will change

NicolasT commented on docs/developer/solutions/deployment.uml:
```diff
@@ -0,0 +1,89 @@
+        |||
+    else ISO is valid and version is present in "MetalK8sSolutions" ConfigMap
+        |||
+        saltmaster -> saltmaster : Skip
```
Why skip? Should be a no-op, so why not just run it?

gdemonet replied:
I'll remove this distinction.

NicolasT commented on docs/developer/solutions/introduction.rst:
```diff
@@ -0,0 +1,137 @@
+
+.. note::
+
+   Removing an archive path from the ``solutions`` list will effectively
```
This is not reflected in the UML diagram I believe?

gdemonet replied:
Indeed, I forgot to add it.

NicolasT commented on docs/developer/solutions/operator.rst:
```diff
@@ -0,0 +1,96 @@
+archive. Those files should be organized as follows::
+
+   operator
+   ├── crd
```
Why not have `crds` (plural) inside `deploy`, as is done by `operator-sdk`?

gdemonet replied:
Because CRDs are deployed differently than the rest, but agreed, it would make sense to keep the whole structure as generated by Operator SDK.

NicolasT replied:
Their deployment is indeed distinct, but I wouldn't translate into the ISO layout, and keep things the way Operator SDK generates them indeed.

NicolasT commented on docs/developer/solutions/operator.rst:
```diff
@@ -0,0 +1,96 @@
+- Operators should create ``Service`` and ``ServiceMonitor`` objects for each
+  deployed component they own
+
+The `Prometheus Operator`_ deployed by MetalK8s will have cluster-scoped
```
`has`, not `will have`.

NicolasT commented on docs/developer/solutions/operator.rst:
```diff
@@ -0,0 +1,96 @@
+  deployed component they own
+
+The `Prometheus Operator`_ deployed by MetalK8s will have cluster-scoped
+permissions, and will be able to read the aforementioned ``ServiceMonitor``
```
`is`

slaperche-scality commented on docs/developer/solutions/archive.rst:
```diff
@@ -0,0 +1,177 @@
+Here is an example invocation of the common Unix mkisofs_ tool to generate such
+archive::
+
+   mkisofs
```
I was thinking, for this (and how to generate images with `skopeo`/`hardlink`), would it be worth to extract those "functions" of our build system into some sort of tools/library/whatever that then could be used by both our buildchain AND the solutions?

That would avoid the "reverse-engineer what our buildchain does, put it in the doc and take care of keeping both of them in sync" approach.

gdemonet replied:
Yes, providing shared tooling would be great. I think we'll know more what are the requirements once we start working on a minimal Solution implementation (#1060).

NicolasT commented on docs/developer/solutions/archive.rst:
```diff
@@ -0,0 +1,177 @@
+   NAME=Example
+   VERSION=1.0.0-dev
+   REQUIRE_METALK8S=">=2.0"
+
```
One thing must be added here: some kind of 'format version', such that we can change the ISO layout over time, and dispatch accordingly during ISO import (in the MetalK8s code).

gdemonet committed:
Fix README references

gdemonet committed:
Recommend build-time information in product.txt

We suggest Solution archives contain a Git reference (if applicable)
and a timestamp in their product information.

Issue: #952

gdemonet committed:
Explain how to generate NGINX configuration

This does not need to be generated by MetalK8s, simply templated.
We provide the required options to pass to the
`static-container-registry.py` script.

Issue: #952

gdemonet committed:
Fix layout for Solution Operator

We want to keep it as close to what Operator SDK generates.

Issue: #952
Issue: #953

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @jbertran

* @nootal

* @cmonfort

* @lucieleonard

* @alxf

* @carlito-scality

* @alexandre-allard-scality

* @teddyandrieux

* @slaperche-scality

* @thomasdanan

* @zempashi

* @nicolast

* @sayf-eddine-scality

* @monpote

* @gdemonet

* @tcarmet





The following reviewers are expecting changes from the author, or must review again:

* @nicolast

bert-e commented:
# Branches have diverged



This pull request's source branch `feature/951-solutions-base-docs` has diverged from
`development/2.4` by more than 50 commits.

To avoid any integration risks, please re-synchronize them using one of the
following solutions:

* Merge `origin/development/2.4` into `feature/951-solutions-base-docs`
* Rebase `feature/951-solutions-base-docs` onto `origin/development/2.4`

Note: If you choose to rebase, you may have to ask me to rebuild
integration branches using the `reset` command.

gdemonet commented:
@bert-e approve

bert-e commented:
# Branches have diverged



This pull request's source branch `feature/951-solutions-base-docs` has diverged from
`development/2.4` by more than 50 commits.

To avoid any integration risks, please re-synchronize them using one of the
following solutions:

* Merge `origin/development/2.4` into `feature/951-solutions-base-docs`
* Rebase `feature/951-solutions-base-docs` onto `origin/development/2.4`

Note: If you choose to rebase, you may have to ask me to rebuild
integration branches using the `reset` command.




*The following options are set:* **approve**

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @jbertran

* @nootal

* @cmonfort

* @lucieleonard

* @alxf

* @carlito-scality

* @alexandre-allard-scality

* @teddyandrieux

* @slaperche-scality

* @thomasdanan

* @zempashi

* @nicolast

* @sayf-eddine-scality

* @monpote

* @gdemonet

* @tcarmet










*The following options are set:* **approve**

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/2.4`



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/2.4`



The following branches have **NOT** changed:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`



Please check the status of the associated issue None.

Goodbye gdemonet.
